rfcs/2020-08-31-3645-graphql-api.md

```diff
@@ -0,0 +1,279 @@
+
+I anticipate doing something similar with the Vector API. async-graphql has a [Context](https://async-graphql.github.io/async-graphql/en/context.html) struct, which is typically used for passing in shared resources such as a database connection pool, or request-specific data such as the current user session.
+
+This is largely TBD, but the basic mechanisms are there to allow for flexible auth when we need it.
+
+## Doc-level Proposal
+
+A Vector observability layer has already been agreed internally. Work is underway.
+
+This proposal discusses the protocol which will govern communications between a running Vector instance, and a web UI and CLI.
+
+The result of this proposal won't directly impact user interaction with observability tooling.
+
+## Rationale
+
+We have built a decent body of experience with GraphQL at Timber, albeit on non-Vector projects.
+
+The tooling I am proposing here represents the same stack (save for swapping Apollo for Urql.)
+
+In a previous Timber project, we used [gqlgen](https://github.com/99designs/gqlgen) as our server library, written in Go. The tooling with async-graphql is similar, though Rust's language features enable a more composible 'code first' approach using macros.
+
+I believe the type of data we are consuming benefits from a strongly typed interface, and that compile-time client generation will significantly reduce the time-to-market of what will be already be a very complex front-end web app.
+
+## Prior Art
+
+See above
+
+## Drawbacks
+
+* GraphQL is generally geared toward graph data (hence the name), although this isn't a hard requirement. IMO, it's a solid general-purpose API interface that mimics what we'd generally wind up creating internally anyway.
+* Queries with unknown levels of nesting can't be queried statically; generally, a need for queries with flatter structures arise.
+* Versioning is trickier, given the lack of distinct endpoints. Nothing stopping multiple endpoints representing versions, although typically this is done in the schema.
+* Some annoying rough edges -- the lack of input unions, for example, typically worked around with more specific queries.
+* Reliance on a relatively young library rather than a 'pure' client like a stdlib tcp server.
+* A need to invest, longer-term, in understanding and contributing to an external library.
+
+## Alternatives
+
+Essentially any browser-compatible protocol could be used, and any text/binary format. This could include other libs such as Protobuf, which might form of a hybrid of the aforementioned approaches.
+
+The maturity of tooling would need further investigation, as well as the development experience of working with any given format.
+
+## Outstanding Questions
+
+* What are the performance characteristics we care about?
+* How many clients do we assume we will be serving across the total of all `vector top` and Vector UI requests?
+* Is JSON an acceptable format for data exchange?
```

Comment:
Is JSON usually exposed directly by GraphQL client libraries or is it generally a hidden implementation detail?

Is there support for compression on the WebSocket streams? JSON can be wordy, but compression usually handles that just fine even for high-volume workloads.

Reply:
> Is JSON usually exposed directly by GraphQL client libraries or is it generally a hidden implementation detail?

Most GraphQL clients deserialize the JSON payload into something meaningful to the language. In Javascript, the payload becomes an object with exactly the same shape. In C# clients, I've seen this be an instance of a class, or an anonymous record.

Whether the client chooses to make the raw JSON payload available is up to the client; in [urql](https://formidable.com/open-source/urql/), we can use middleware to get the JSON, or the JS stdlib to turn it back into JSON:

```js
JSON.stringify(data, null, 2); // <-- JSON with formatting
```

I do this in the [development workflow video](https://www.loom.com/share/9ca38feb21bb488f92d729df0148e029) to show a JSON representation in React.

> Is there support for compression on the WebSocket streams? JSON can be wordy, but compression usually handles that just fine even for high-volume workloads.

There's `permessage-deflate` and custom extensions available to compress JSON payloads over WS if we feel we need it. Payloads are typically a couple of kb or less, depending on the query; we'd have to jam a decent volume of them down the wire before it affected the typical end-client.

async-graphql's `graphql_subscription` middleware is a thin wrapper around Warp's WS connection handling:

```rust
 warp::any()
        .and(warp::ws())
        .and(warp::any().map(move || schema.clone()))
        .map(
            |ws: warp::ws::Ws, schema: Schema<Query, Mutation, Subscription>| {
                ws.on_upgrade(move |websocket| {
                    let (mut tx, rx) = websocket.split();
                    let (mut stx, srx) =
                        schema.subscription_connection(WebSocketTransport::default());

                    let mut rx = rx.fuse();
                    let mut srx = srx.fuse();

                    async move {
                        loop {
                            select! {
                                bytes = srx.next() => {
                                    if let Some(bytes) = bytes {
                                        if let Ok(text) = String::from_utf8(bytes.to_vec()) {
                                            if tx.send(Message::text(text)).await.is_err()
                                            {
                                                return;
                                            }
                                        }
                                    } else {
                                        return;
                                    }
                                }
                                msg = rx.next() => {
                                    if let Some(Ok(msg)) = msg {
                                        if msg.is_text() {
                                            if stx.send(Bytes::copy_from_slice(msg.as_bytes())).await.is_err() {
                                                return;
                                            }
                                        }
                                    } else {
                                        return;
                                    }
                                }
                            }
                        }
                    }
                })
            },
        ).map(|reply| {
        warp::reply::with_header(reply, "Sec-WebSocket-Protocol", "graphql-ws")
    })
        .boxed()
```

What we send to the `tx` channel is up to us; we could swap this out with compressed bytes, or any other format.

The caveat is that the payload is then a DIY implementation. A connecting client would need to be aware of that divergence from the norm, and interpret accordingly. I think a more practical implementation would be passing  headers assert compression capabilities, and selectively compressing (as is typical with `Accept: br`, etc), but I'd consider that a tweak we can kick down the road until we know it's actually warranted.

Warp might even have some helpers there. I know it negotiates Brotli and Gzip for standard HTTP headers, but don't know if it offers anything special for WS. We can always roll our own.